clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java

```diff
@@ -109,101 +132,89 @@ public void run() {
-                    log.debug(
-                        "Exception thrown, background thread won't terminate",
-                        e
-                    );
-                    // swallow the wakeup exception to prevent killing the
-                    // background thread.
+                    log.debug("WakeupException caught, background thread won't be interrupted");
+                    // swallow the wakeup exception to prevent killing the background thread.
                 }
             }
         } catch (final Throwable t) {
-            log.error(
-                "The background thread failed due to unexpected error",
-                t
-            );
-            if (t instanceof RuntimeException)
-                this.exception.set(Optional.of((RuntimeException) t));
-            else
-                this.exception.set(Optional.of(new RuntimeException(t)));
+            log.error("The background thread failed due to unexpected error", t);
+            throw new RuntimeException(t);
         } finally {
             close();
             log.debug("{} closed", getClass());
         }
     }
 
     /**
-     * Process event from a single poll
+     * Poll and process an {@link ApplicationEvent}. It performs the following tasks:
+     *  1. Try to poll and event from the queue, and try to process it using the coorsponding {@link ApplicationEventProcessor}.
+     *  2. Try to find Coordinator if needed
+     *  3. Poll the networkClient for outstanding requests.
      */
     void runOnce() {
-        this.inflightEvent = maybePollEvent();
-        if (this.inflightEvent.isPresent()) {
-            log.debug("processing application event: {}", this.inflightEvent);
+        Optional<ApplicationEvent> event = maybePollEvent();
+
+        if (event.isPresent()) {
+            log.debug("processing application event: {}", event);
+            consumeApplicationEvent(event.get());
         }
-        if (this.inflightEvent.isPresent() && maybeConsumeInflightEvent(this.inflightEvent.get())) {
-            // clear inflight event upon successful consumption
-            this.inflightEvent = Optional.empty();
+
+        final long currentTimeMs = time.milliseconds();
+        // TODO: This is just a place holder value.
+        long pollWaitTimeMs = 100;
+
+        // TODO: Add a condition here, like shouldFindCoordinator in the future.  Since we don't always need to find
+        //  the coordinator.
+        if (coordinatorManager.isPresent()) {
+            pollWaitTimeMs = Math.min(pollWaitTimeMs, handlePollResult(coordinatorManager.get().poll(currentTimeMs)));
         }
 
         // if there are pending events to process, poll then continue without
         // blocking.
-        if (!applicationEventQueue.isEmpty() || inflightEvent.isPresent()) {
-            networkClient.poll(time.timer(0));
+        if (!applicationEventQueue.isEmpty()) {
+            networkClientDelegate.poll(0);
```

Comment:
It may take a little while before we are able to send a request (e.g. the connection might have reached the max inflight requests allowed). We do not want to busy loop in this case. I'd suggest using `retryBackoffMs` in this case. We can structure the logic like this.

```java
if (!applicationEventQueue.isEmpty()) {
  pollWaitTimeMs = Math.min(retryBackoffMs, pollWaitTimeMs);
}

networkClientDelegate.poll(pollWaitTimeMs);
```

Reply:
I had the context a little wrong here. I wonder if we need this at all. Instead, we should try to drain all application events before we poll the network client. And when we post a new event, we should use `wakeup()`.